Better Touchscreen Zooming

personalizedrefrigerator opened the pull request:
## Addresses these issues:
 * Previously, when touch drawing was enabled, the touchscreen could not be used to pan and zoom. 
 * Additionally, zoom with touch caused the page to scroll up and down by large amounts (zoom centered at top-left of document).
      * **Edit:** This only happened when touch drawing/the touchscreen workaround was enabled.
 * Zooming can be very slow when editing large PDFs.
      * **Only partially fixes this issue** 

## Adds these features:
  * Two-finger touch panning/scrolling when touch drawing is enabled.
  * An option to clear the PDF page cache less frequently.
  * Pressure inference (faster strokes -> lower pressure).
      * An option to toggle this in settings.

## Screenshot
![image](https://user-images.githubusercontent.com/46334387/100393728-84678c80-2fef-11eb-871a-57846467f412.png)


## Questions I Have:
 * What is the reason for the workaround that enables touch drawing? Is it reasonable to allow panning and zooming when the user uses multi-touch with this workaround enabled?

LittleHuba commented on src/gui/inputdevices/TouchInputHandler.cpp:
```diff
@@ -119,8 +129,9 @@ void TouchInputHandler::zoomStart() {
     }
 
     Rectangle zoomSequenceRectangle = zoomControl->getVisibleRect();
+    auto topLeft = utl::Point<double>{zoomSequenceRectangle.x, zoomSequenceRectangle.y};
```
Please merge this with `center` as it does not make sense to have this variable declared as center if it actually isn't.

personalizedrefrigerator replied:
Done! Interestingly, however, this breaks zooming when touch drawing/the workaround is disabled.

For now, I change the center here only if the workaround/touch zooming is enabled.

LittleHuba commented on src/gui/inputdevices/PenInputHandler.cpp:
```diff
@@ -142,13 +142,33 @@ auto PenInputHandler::actionStart(InputEvent const& event) -> bool {
     return false;
 }
 
+double PenInputHandler::inferPressure(PositionInputData const& pos, XojPageView* page) {
```
Two things with infering pressure. I generally like the idea behind it, as drawing quickly is often associated with less pressure. But not all users might like this as it could be understood as unintuitive. Hence I'd appreciate it if you could make this optional via a preference in the preferences dialog.

Additionally this feature is quite specific to the TouchDrawingInputHandler and should therefore reside in this class. If you add it to the PenInputHandler, you would block any new tool that might actually not want pressure by design.

personalizedrefrigerator replied:
I think it might make sense for pressure to be enabled for multiple input sources. For example, mice also lack pressure sensitivity.

I have added this to the preferences dialog.

LittleHuba reviewed (changes requested):
Hi!
Thanks for the work! This are actually some amazing additions. I could not actually test the features so far, but got around to at least give your code a once-over.

LittleHuba commented:
> What is the reason for the workaround that enables touch drawing? Is it reasonable to allow panning and zooming when the user uses multi-touch with this workaround enabled?

The workaround you can specifically select in the preferences is a feature of the old input system. The new system does not require nor use this feature. Instead it uses the list of devices where you can assign device classes. So if you assign a touchscreen the pen input class it will actually draw for touch inputs. This is interesting for tablets where you don't actually have a stylus to draw with. In that manner it would also be interesting to enable panning and zooming for multi-touch gestures. But it should be optional and not interfere with the response of the normal inputs. You'll have to do some magic for this where you treat the input initially as normal input and the invalidate the input when the second finger touches the screen.

You can text me on Gitter or Matrix if you want some more details or didn't understand what I tried to describe here.

personalizedrefrigerator commented:
I updated the preferences dialog! I, however, have not updated any of the translations...

personalizedrefrigerator commented:
@LittleHuba I think I've made the requested changes!

There are still some issues, however:
 * When touch drawing is enabled, writing outside the window doesn't work.
 * New settings in the SettingsDialog lack translations.

Febbe commented on src/gui/MainWindow.cpp:
```diff
@@ -282,15 +284,44 @@ void MainWindow::initXournalWidget() {
+
+    // Run on the main thread:
+    // 0 = timeout (run as soon as possible)
+    gdk_threads_add_timeout(
```
**nitpick (non-blocking):** use `gdk_threads_add_idle*` here instead of `gdk_threads_add_timeout`

personalizedrefrigerator replied:
I tried `gtk_threads_add_idle(`, but the priority it put calling the function at too low of a priority!

I'm looking into using `gtk_threads_add_idle_full` with a priority of `G_PRIORITY_HIGH_IDLE`!

personalizedrefrigerator replied:
According to the GTK documentation, `gtk_threads_add_idle_full` has the following signature:
```
guint
gdk_threads_add_idle_full (gint priority,
                           GSourceFunc function,
                           gpointer data,
                           GDestroyNotify notify);
```

Personally, `gdk_threads_add_timeout` seems more readable here; however, I understand that this might not be a question of readability.

Which do you think is more readable/better?

Febbe replied:
It is a question of understandability and this is a subset of readability. `gtk_threads_add_idle` looks more like adding a task to the scheduler. A timeout looks like, that the code must be executed after a specific amount of time with a specific frequency. But also with your above comment this should be the same.
I leave that to you and @LittleHuba.

Regarding my opinion: `gtk_threads_add_idle` is more readable, since it also tells the reader which priority must be used. Also, the `GDestroyNotify` can be a `nullptr` as I know. This would only add one parameter and will replace the magic number `0` with a named constant.

personalizedrefrigerator replied:
I found `Util::execInUiThread`! I'm adding a `priority` parameter and using that!

LittleHuba replied:
`Util::execInUiThread` is actually meant for exactly this usecase.

personalizedrefrigerator replied:
Changed!

Febbe commented on src/gui/MainWindow.cpp:
```diff
@@ -282,15 +284,44 @@ void MainWindow::initXournalWidget() {
+                gtk_scrolled_window_set_kinetic_scrolling(GTK_SCROLLED_WINDOW(self->winXournal),
+                                                          self->gtkTouchscreenScrollingEnabled);
+
+                return FALSE;
```
Macros are discouraged since c++11 and mostly obsolete since c++17. Please use the in build types.

```suggestion
                return false;
```

Febbe commented on src/gui/inputdevices/TouchInputHandler.cpp:
```diff
@@ -118,12 +128,22 @@ void TouchInputHandler::zoomStart() {
 }
 
 void TouchInputHandler::zoomMotion(InputEvent const& event) {
+    if (!inputContext->getSettings()->isZoomGesturesEnabled()) {
+        return;
+    }
```
**Discussion:** This looks like a workaround for me, isn't it. In my opinion this function or this InputHandler shouldn't be called in that case.

personalizedrefrigerator replied:
I was duplicating logic found at the beginning of `zoomStart`! 

I can change this!

personalizedrefrigerator replied:
Changed!

Febbe commented on src/gui/inputdevices/TouchInputHandler.cpp:
```diff
@@ -142,10 +162,13 @@ void TouchInputHandler::zoomMotion(InputEvent const& event) {
 }
 
 void TouchInputHandler::zoomEnd() {
+    if (!inputContext->getSettings()->isZoomGesturesEnabled()) {
+        return;
+    }
+
```
**Discussion:** same here

personalizedrefrigerator commented on src/control/PdfCache.cpp:
```diff
@@ -46,53 +45,63 @@ void PdfCache::clearCache() {
         this->data.pop_back();
     }
+
+    auto* ne = new PdfCacheEntry(std::move(popplerPage), img, zoom);
+    this->data.push_front(ne);
+
+    return ne;
```
The change here was to make `PdfCache::cache` return a `PdfCacheEntry*` and cache the `zoom`... The `new` isn't new!

The surrounding diff looks like this:
```
-  void PdfCache::cache(XojPdfPageSPtr popplerPage, cairo_surface_t* img) {
-  auto* ne = new PdfCacheEntry(std::move(popplerPage), img);
-    this->data.push_front(ne);
-
-    while (this->data.size() > this->size) {
+ PdfCacheEntry* PdfCache::cache(XojPdfPageSPtr popplerPage, cairo_surface_t* img, double zoom) {
+    while (this->data.size() >= this->size) {
+        delete this->data.back();
+        this->data.pop_back();
+    }
+
+    auto* ne = new PdfCacheEntry(std::move(popplerPage), img, zoom);
+    this->data.push_front(ne);
+
+    return ne;
```

I agree with you, though!!! Refactoring would be nice here!!!

personalizedrefrigerator replied:
I'm resolving this issue. If you have further comments, feel free to un-resolve it!

Febbe replied:
No, I didn't see that, this should be resolved in another PR.
With that the `std::list<PdfCacheEntry*>` should also be replaced by `std::deque<PdfCacheEntry>`

personalizedrefrigerator commented on src/gui/MainWindow.h:
```diff
@@ -165,6 +167,9 @@ class MainWindow: public GladeGui, public LayerCtrlListener {
     GtkWidget* winXournal = nullptr;
     ScrollHandling* scrollHandling = nullptr;
 
+    bool usingTouchWorkaround;
+    bool gtkTouchscreenScrollingEnabled;
```
According to its description in the settings dialog, the touchscreen workaround should only take effect when the application restarts. 

`usingTouchWorkaround` is cached at program run to provide this functionality -- `settings->isTouchWorkaround()` reflects the current state of the setting. It might make more sense for `settings->isTouchWorkaround()` to return the current **effective state** of `touchWorkaroundEnabled`. Unfortunately, the preferences dialog uses this to set the state of the touch workaround checkbox... 

Similarly, `gtkTouchscreenScrollingEnabled` is necessary as (I think) it isn't stored anywhere else (except GTK).

Febbe commented:
@LittleHuba I may want to take a closer look at the zoom changes in `TouchInputHandler.cpp`, since I rewrote parts of the zooming handling. 
@personalizedrefrigerator that's the only reason, why I also made a small first review. Mostly nitpicks and suggestions. Your code looks very well and I like your pressure inference feature. Keep going :).

LittleHuba commented:
@Febbe of course go ahead with that part. You most likely know that system way better than me.

LittleHuba commented on src/gui/inputdevices/PenInputHandler.cpp:
```diff
@@ -142,13 +143,33 @@ auto PenInputHandler::actionStart(InputEvent const& event) -> bool {
+        double dt = std::min((pos.timestamp - lastPos.timestamp) / 10.0, 2.0);
+        double inverseSpeed = (dt) / std::sqrt(std::pow(pos.x - lastPos.x, 2) + std::pow(pos.y - lastPos.y, 2) + 0.001);
+
+        double newPressure = 3.142 / 2.0 + std::atan(inverseSpeed * 3.14 - 1.3);
```
You are using very expensive methods here (pow, sqrt, atan) within a function that is called very very often. It would be good to simplify this calculation to a quick approximation. The result does not need to be perfect as it has no huge influence on the look if the pressure is 0.00000002 or 0.00000003.

Febbe replied:
`std::pow(x,2)` can be replaced by x*x, which is much faster. Unfortunately there is no constexpr version of `<cmath>` functions. Then it wouldn't make any difference.
Also `<cmath>` supports the [`std::hypot`](https://en.cppreference.com/w/cpp/numeric/math/hypot) function, which is optimized for calculating the 2nd norm. If you require precision instead of a numeric approximation, it is still a way faster than a manual `sqrt(pow(x,2) + pow(y,2))` (in this case just use `utl::Point::distance` or `Point::lineLengthTo).

personalizedrefrigerator replied:
`pos` and `lastPos` are of type `PositionInputData`. As such, `utl::Point::distance` does not apply without a conversion.

I am considering making `PositionInputData` extend `utl::Point<double>`...

LittleHuba replied:
I think extending `util::Point<double>` would be a valid solution.

personalizedrefrigerator replied:
I'm still not sure what to do about `std::atan`...

Febbe replied:
You could use a local object of type utl::Point<double>. Most likely this will completely inlined by the compiler :).

personalizedrefrigerator replied:
Extending `utl::Point<double>` makes the code much cleaner. I think I'll do that for now...

LittleHuba reviewed (changes requested):
Nice work so far. What I'm missing in general are comments. I know they are a huge bother to write every time but it just makes the work for me and the next one working on this code so much easier.

personalizedrefrigerator commented:
@Febbe Most of the changes to `TouchInputHandler.cpp` were an attempt to fix view translation while scrolling large documents:

![out](https://user-images.githubusercontent.com/46334387/100444669-aa794500-3060-11eb-9f78-ad403622bb37.gif)

I think this is because the distance between pages is constant, even while zooming. As such, zooming out increases the **relative** distance between pages. Neither `ZoomControl` nor `Layout` take this into account. 

As such, zooming does not work expected near the end of many-page documents.

personalizedrefrigerator commented on src/control/PdfCache.cpp:
```diff
@@ -46,53 +45,63 @@ void PdfCache::clearCache() {
-
-    while (this->data.size() > this->size) {
+PdfCacheEntry* PdfCache::cache(XojPdfPageSPtr popplerPage, cairo_surface_t* img, double zoom) {
+    while (this->data.size() >= this->size) {
```
Oops! I must have changed this while debugging! I can change it back!

personalizedrefrigerator replied:
Reverted!

Febbe commented on src/control/ToolHandler.cpp:
```diff
@@ -152,9 +154,15 @@ void ToolHandler::selectTool(ToolType type, bool fireToolChanged, bool triggered
 void ToolHandler::fireToolChanged() {
-    if (listener) {
-        listener->toolChanged();
+    for (auto listener: this->toolChangeListeners) {
```
This copies all listeners in `this->toolChangeListeners`. Was this intentional?
Better use:
```suggestion
    for (auto&& listener: this->toolChangeListeners) {
```

personalizedrefrigerator replied:
No!!!! Thank you!!!!

Febbe commented on src/control/ToolHandler.cpp:
```diff
@@ -152,9 +154,15 @@ void ToolHandler::selectTool(ToolType type, bool fireToolChanged, bool triggered
+}
+
+void ToolHandler::addToolChangedListener(std::function<void(ToolType)> listener) {
+    toolChangeListeners.push_back(listener);
```
```suggestion
    toolChangeListeners.emplace_back(std::move(listener));
```

Febbe commented on src/control/ToolHandler.h:
```diff
@@ -41,7 +42,7 @@ class ActionHandler;
 class ToolHandler {
 public:
-    ToolHandler(ToolListener* listener, ActionHandler* actionHandler, Settings* settings);
+    ToolHandler(ToolListener* stateChangedListener, ActionHandler* actionHandler, Settings* settings);
```
```suggestion
    using ToolChangedCallback = std::function<void(ToolType)>;
    ToolHandler(ToolListener* stateChangedListener, ActionHandler* actionHandler, Settings* settings);
```

Febbe commented on src/control/ToolHandler.h:
```diff
@@ -93,6 +94,10 @@ class ToolHandler {
 
+    // Different from the listener given to the constructor -- [listener]
+    // here only listens for when the current tool is changed to another.
+    void addToolChangedListener(std::function<void(ToolType)> listener);
```
```suggestion
    void addToolChangedListener(ToolChangedCallback listener);
```

Febbe commented on src/control/ToolHandler.h:
```diff
@@ -149,7 +154,8 @@ class ToolHandler {
      */
 
-    ToolListener* listener = nullptr;
+    std::vector<std::function<void(ToolType)> > toolChangeListeners;
```
```suggestion
    std::vector<ToolChangedCallback> toolChangeListeners;
```

personalizedrefrigerator commented on src/gui/inputdevices/TouchDrawingInputHandler.cpp:
```diff
@@ -96,5 +156,18 @@ auto TouchDrawingInputHandler::changeTool(InputEvent const& event) -> bool {
+    if (mainWindow != nullptr && control->getSettings()->getTouchDrawingEnabled()) {
+        mainWindow->setGtkTouchscreenScrollingEnabled(toolHandler->getToolType() == TOOL_HAND);
+    }
+}
```
Fixed!

LittleHuba commented:
> @LittleHuba I think I've made the requested changes!
> 
> There are still some issues, however:
> 
> * When touch drawing is enabled, writing outside the window doesn't work. **Edit:** This has been fixed.
> * New settings in the SettingsDialog lack translations.
> * (**Edit**) `Recent Documents` isn't working. **This might have been broken before. Please test in `xournalpp:master` before merging.**

The missing translations will be fixed by the community in time. If there exists no translation, the english tag is used. People can introduce new translations easily using our Crowdin project.

Recent documents was bugged some time ago. I'm not quite sure if the fix is already merged. But that should not be of concern for this PR as it is a completely different topic.

Febbe commented on src/control/ToolHandler.cpp:
```diff
@@ -12,11 +12,13 @@
     this->settings = settings;
     initTools();
-    this->listener = listener;
+    this->toolChangeListeners = std::vector<std::function<void(ToolType)> >();
```
```suggestion
```

Febbe commented on src/control/ToolHandler.cpp:
```diff
@@ -152,9 +154,15 @@ void ToolHandler::selectTool(ToolType type, bool fireToolChanged, bool triggered
+    stateChangeListener->toolChanged();
+}
+
+void ToolHandler::addToolChangedListener(std::function<void(ToolType)> listener) {
```
```suggestion
void ToolHandler::addToolChangedListener(ToolChangedCallback listener) {
```

LittleHuba commented:
> @Febbe Most of the changes to `TouchInputHandler.cpp` were an **attempt** to fix view translation while scrolling large documents:
> 
> ![out](https://user-images.githubusercontent.com/46334387/100444669-aa794500-3060-11eb-9f78-ad403622bb37.gif)
> 
> I think this is because the distance between pages is constant, even while zooming. As such, zooming out increases the **relative** distance between pages. Neither `ZoomControl` nor `Layout` take this into account.
> 
> As such, zooming does not work expected near the end of many-page documents.
> 
> ### This is still an issue.

Would you be interested in doing a follow-up PR that fixes this issue?

personalizedrefrigerator commented:
> 
> Would you be interested in doing a follow-up PR that fixes this issue?

Yes. This issue has been bothering me. I can try to fix it.

LittleHuba commented:
We'll either have to squash the commits before merging the PR or you do some rebasing-magic to clean up your commits. Currently it is a lot of commits just fixing previous errors.

personalizedrefrigerator commented:
> We'll either have to squash the commits before merging the PR or you do some rebasing-magic to clean up your commits. Currently it is a lot of commits just fixing previous errors.

I'm working on this!

Febbe commented on src/gui/dialog/SettingsDialog.h:
```diff
@@ -35,9 +35,15 @@ class SettingsDialog: public GladeGui {
     void enableWithCheckbox(const string& checkbox, const string& widget);
+    void disableWithCheckbox(const string& checkbox, const string& widget);
+
+    /**
```
**Nitpick (non-blocking):** I would not use `/**`, since this marks the comment as documentation for the next function and will be associated with it. Since the content is a description for the whole block, we do not want this.

```suggestion
    /*
```

Febbe commented on src/gui/inputdevices/PenInputHandler.cpp:
```diff
@@ -142,13 +143,33 @@ auto PenInputHandler::actionStart(InputEvent const& event) -> bool {
+        PositionInputData lastPos = getInputDataRelativeToCurrentPage(page, this->lastEvent);
+
+        double dt = std::min((pos.timestamp - lastPos.timestamp) / 10.0, 2.0);
+        double inverseSpeed = (dt) / (pos.distance(lastPos) + 0.001);
```
```suggestion
        double inverseSpeed = (dt) / (utl::Point<double>(pos.x,pos.y).distance(utl::Point<double>(lastPos.x,lastPos.y)) + 0.001);
```

Febbe commented on src/gui/inputdevices/PositionInputData.h:
```diff
@@ -14,7 +14,9 @@
-class PositionInputData {
+#include "util/Point.h"
+
+class PositionInputData: public utl::Point<double> {
```
Do not use inheritance here:
 1. `utl::Point<>` does not provide a virtual destructor, which will introduce unnecessarily a v-table for such a trivial struct.
 2. `PositionInputData` is not a Point it contains one. So make `utl::Point<double>` a member with the name `pos` or `position`. A doc is not required, since this is self explaining.

LittleHuba commented on po/README.md:
```diff
@@ -0,0 +1,3 @@
+Please use [Crowdin](https://crowdin.com/project/xournalpp) to create or edit translations. 
+Should a language not be available, you can reach out to the developers by opening a new 
+issue or Gitter/Matrix (see links in main README).
```
Looks like one of your rebases was not correct. This change is from a different PR and should not be part of your changes.

personalizedrefrigerator replied:
I merged changes to `xournalpp:master` into my branch partway through work on my PR. I probably shouldn't have done this.

![image](https://user-images.githubusercontent.com/46334387/100514298-4f128a00-3128-11eb-8e6a-39ae3a3420a6.png)

These changes probably come from the commits labeled "**LittleHubba** authored and **personalizedrefrigerator committed**".

personalizedrefrigerator replied:
I may have to fork `xournalpp:master` again and apply my changes to the fork... Should I do this and open a new PR?

LittleHuba replied:
Normally you do not merge both branches but rebase on top of master.
You can use the git reflog to undo the merge and then do the rebase.

personalizedrefrigerator replied:
Okay. Thank you for catching this!

LittleHuba replied:
If using the reflog does not work, just cherry pick your changes to another branch locally and force push to your PR branch.

Febbe replied:
Yes instead of using `git merge upstream:master` one should ever use `git rebase upstream:master` this will keep the history clean and you won't mess with foreign commits.

personalizedrefrigerator commented:
As it relates to the topic of this PR, I have included a **partial** fix for the moving-viewport-while-zooming issue (I think issue #2411). See [this commit](https://github.com/xournalpp/xournalpp/pull/2435/commits/550375a3bb71c6f3b052be7c5ab251154608117d).

LittleHuba commented on src/control/zoom/ZoomControl.cpp:
```diff
@@ -78,15 +79,26 @@ auto ZoomControl::getVisibleRect() -> Rectangle<double> {
     }
 
-    return (this->scrollPosition * this->zoom) - this->zoomWidgetPos;
+    return (this->scrollPosition * this->zoom) - this->zoomWidgetPos + this->unscaledPixels;
```
Not a very good idea to use the variable for this. This forces the call of `void ZoomControl::setScrollPositionAfterZoom(utl::Point<double> scrollPos)` before any call to `ZoomControl::getScrollPositionAfterZoom()` may happen or it returns the wrong result. As getPaddingAbovePage is also const you should just call the method here as well.

personalizedrefrigerator replied:
Notice that `ZoomControl::startZoomSequence` calls `ZoomControl::setScrollPositionAfterZoom`! 

Of course, I can still make this change -- it gets rid of an unnecessary member variable ☺.

personalizedrefrigerator replied:
`ZoomControl::startZoomSequence` calls `ZoomControl::setScrollPositionAfterZoom`. To call `getPaddingAbovePage`, we also need to call `XournalView::getWidget()`, which is not `const` ☺.

personalizedrefrigerator replied:
I should document this!

personalizedrefrigerator replied:
Done!

I should note that this change is only a partial fix. While the zooming in/out very far into a large document no longer causes the viewport to scroll significantly, it does cause the viewport to rapidly move up and down small amounts.

For anyone working on this in the future: Having set `XOURNAL_PADDING_BETWEEN` and tested before making these changes, I saw that this behavior occurs without this new partial fix.

LittleHuba commented on src/gui/sidebar/previews/base/SidebarPreviewBase.cpp:
```diff
@@ -128,6 +128,6 @@ auto SidebarPreviewBase::scrollToPreview(SidebarPreviewBase* sidebar) -> bool {
 }
 
-void SidebarPreviewBase::pageDeleted(int page) {}
+void SidebarPreviewBase::pageDeleted(size_t page) {}
```
For those really large documents with more than 2147483647 pages? 😉 
But leave it there, it doesn't hurt.

personalizedrefrigerator replied:
I was getting a warning in the mac build CI check :)

LittleHuba reviewed (approved):
This seems to be the best fix for now. In the long run we should improve on this. But as we need to revamp the rendering system anyway it does not make sense to put more work in this now.

I'll do a functional test in the next few days and given everything works we can merge this.

personalizedrefrigerator commented:
![image](https://user-images.githubusercontent.com/46334387/101258124-37617580-36d5-11eb-9031-fbad2c0dbba2.png)

The "Pressure Inference" frame lacks a border. This stylistic fix really should be a part of commit 1450a7c923b69055de8f0c677b06b2aec8e4e752. As such, I'm making the change and force-pushing the modified history.

LittleHuba commented:
> ![image](https://user-images.githubusercontent.com/46334387/101258124-37617580-36d5-11eb-9031-fbad2c0dbba2.png)
> 
> The "Pressure Inference" frame lacks a border. This stylistic fix really should be a part of commit 1450a7c923b69055de8f0c677b06b2aec8e4e752 (and only involves changes to a few lines of `settings.glade`). As such, I'm making the change and force-pushing the modified history.

This should actually be part of the previous frame.

personalizedrefrigerator commented:
Already changed!

personalizedrefrigerator commented:
This is what the changed parts of the settings dialog currently looks like:
![image](https://user-images.githubusercontent.com/46334387/101295176-e59e1580-37d0-11eb-8046-592e91d2c4bd.png)
![image](https://user-images.githubusercontent.com/46334387/101295190-f2226e00-37d0-11eb-8207-ccfb7e238235.png)
![image](https://user-images.githubusercontent.com/46334387/101295198-fea6c680-37d0-11eb-8d93-fa8ec4d3cd54.png)
![image](https://user-images.githubusercontent.com/46334387/101295224-2ac24780-37d1-11eb-9d64-877a1dcae3dc.png)

LittleHuba commented:
> ![image](https://user-images.githubusercontent.com/46334387/101295176-e59e1580-37d0-11eb-8046-592e91d2c4bd.png)

This should be part of Drawing Area as it is part of the rendering system and has not really anything to do with loading or saving a file.

personalizedrefrigerator commented:
> This should be part of Drawing Area as it is part of the rendering system and has not really anything to do with loading or saving a file.

This has been changed!

personalizedrefrigerator commented:
![image](https://user-images.githubusercontent.com/46334387/101390240-5ac42600-3877-11eb-8484-f1e988b0bc53.png)
![image](https://user-images.githubusercontent.com/46334387/101390280-6ca5c900-3877-11eb-8f1b-832a8c424e22.png)

personalizedrefrigerator commented:
@Febbe @eliasrg  Can you check whether the same issue fixed by c1edabf2ccaa64606cacf2e29790d7425978ec3c is also fixed here? 
In this PR, multi-touch gestures are used even when touch-gesture zooming is disabled. As such, I would prefer to not ignore all multi-touch gestures when touch-gesture zooming is disabled.

eliasrg commented:
@personalizedrefrigerator Yes, #2509 is also fixed here. In fact, with zoom gestures and touch drawing turned off, it seems to work exactly like c1edabf (just ignoring the second finger). So in principle, I have no objection at all!

However, I had a look at your `TouchInputHandler.cpp` and I think I found a small issue with it. I will make a separate comment.

eliasrg commented on src/gui/inputdevices/TouchInputHandler.cpp:
```diff
@@ -36,22 +38,27 @@ auto TouchInputHandler::handleImpl(InputEvent const& event) -> bool {
 
         if (event.sequence == this->primarySequence) {
-            this->primarySequence = nullptr;
+            // If secondarySequence is nullptr, this sets primarySequence
+            // to nullptr. If it isn't, then it is now the primary sequence!
+            this->primarySequence = this->secondarySequence;
+            this->secondarySequence = nullptr;
+
+            this->priLastAbs = this->secLastAbs;
+            this->priLastRel = this->secLastRel;
```
If I understand this part correctly, I should, if zoom gestures and touch drawing are disabled, be able to scroll with one finger, put a second finger on the screen, release the first finger and continue scrolling with the second finger. But this code doesn't accomplish that, because `secLastAbs` and `secLastRel` are only kept up to date if `zoomGesturesEnabled` is true. I wrote a quick fix for this: https://github.com/eliasrg/xournalpp/commit/b9c50aab62ce58d157b5960f8d9c23dd1f98e7c4 and it does work when I test it. However, this logic is getting a bit confusing/unnatural, so it might be better to restructure the file more drastically...

personalizedrefrigerator replied:
I think `TouchInputHandler::scrollMotion` should already be keeping these up to date:
```cpp
void TouchInputHandler::scrollMotion(InputEvent const& event) {
    // Will only be called if there is a single sequence (zooming handles two sequences)
    auto offset = [&]() {
        auto absolutePoint = utl::Point{event.absoluteX, event.absoluteY};
        if (event.sequence == this->primarySequence) {
            auto offset = absolutePoint - this->priLastAbs;
            this->priLastAbs = absolutePoint;
            return offset;
        } else {
            auto offset = absolutePoint - this->secLastAbs;
            this->secLastAbs = absolutePoint;
            return offset;
        }
    }();
```

I might be misunderstanding this...

eliasrg replied:
But `scrollMotion` is not called for the second finger if there are two fingers on the screen, only for the one actually scrolling.

I think you can see the difference if you compile both versions and try switching fingers while scrolling (zoom gestures and touch drawing turned off)?

personalizedrefrigerator replied:
Yes! With touch-zooming off, I am hitting the issue! Thank you for catching this!

I agree with you about re-structuring the file -- it is getting rather difficult to understand what is going on.

eliasrg replied:
Okay, I tried to restructure this in a nice way, but I encountered strange bugs, and eventually I realised that I had misunderstood how the touch input works. Because it doesn't happen only in this file. If you comment out the whole of `handleImpl` except `return false`, you can still scroll with one finger without problem. But as soon as you put a second finger on the screen, it stops working. This is because there is a [`GtkScrolledWindow`](https://developer.gnome.org/gtk3/stable/GtkScrolledWindow.html) which has built-in support for one-finger touches, and even provides kinetic scrolling (which the code in `scrollMotion` doesn't, as you can check by scrolling after switching fingers in my "quick fix" https://github.com/eliasrg/xournalpp/commit/b9c50aab62ce58d157b5960f8d9c23dd1f98e7c4).

So I don't know why there are two competing systems for scrolling... it seems like a bad idea unless I am missing something. Can anybody explain the reason for this? My feeling is that we should either only handle zooming here and leave scrolling for the `GtkScrolledWindow` (probably the best), or get rid of the scrolled window and handle _all_ scrolling ourselves (more work and we wouldn't have kinetic scrolling, but it'd play nicer with zooming and multitouch).

Does that make sense?

personalizedrefrigerator replied:
Thank you for looking into this!
The "enable touch workaround" setting switches between the two scrolling methods.

personalizedrefrigerator replied:
It seems like a proper fix for the transition from two-finger panning to one-finger scrolling will require some refactoring... 
Do you think it could be appropriate to make this fix the subject of a follow-up PR?

LittleHuba replied:
> So I don't know why there are two competing systems for scrolling... it seems like a bad idea unless I am missing something. Can anybody explain the reason for this?

@eliasrg This is actually a fix for a very old problem that was ported to the new input system. When you enable touch drawing (the touch workaround) you were essentially unable to scroll at all. Back in the day this was fixed with this convoluted system of handling scrolling manually in this case. As you mentioned GTK actually would handle this for us but only for "single finger touches" which are used for drawing in the workaround.

I see two solutions here:
1. We stick with this convoluted system and fix it as well as possible. This would keep our code somewhat smaller and leave stuff like kinetic scrolling with GTK. Arguably it would make the input system harder to understand but also minimize potential areas where we need to fix stuff.
2. We throw out GTK scroll handling completely and implement stuff like kinetic scrolling in our own handlers. I would then make it optional if scrolling should be done with two fingers or one (one being default as it is consistent with GTK). This would increase our code size but make the system straight forward.

@Febbe, @idotobi, @rolandlo, @Technius, @personalizedrefrigerator and @eliasrg what is your opinion on this? Lets make a poll with reaction smileys (🎉 for 1 and 🚀 for 2)

eliasrg replied:
@LittleHuba If we are okay with not having two-finger scrolling (personally, I don't need it, but if you are writing with your finger you probably do), then I'd say let GTK do all scroll handling and only handle zoom ourselves. If we do need two-finger scrolling, it seems we've outgrown GTK's scrolling and should throw it out! Since we already have a working scroll handler (which I am working on simplifying), the most difficult part to implement would probably be kinetic scrolling, if indeed we need that. We'd also get a slightly nicer feel overall, with smooth transitions between scrolling and zooming.

LittleHuba replied:
Okay the verdict is quite clear. Well go forward with option two. As this PR is already big enough I'd recommend a follow-up PR for this. @personalizedrefrigerator and @eliasrg would you like to join up and work on this?

LittleHuba commented:
After merging #2339 we got some conflicting files. Could you clean those up? I'd like to merge your PR as one of the next few so it will be included in the next release.

LittleHuba commented:
With the scrolling question in the comment handled I assume this is good to merge @personalizedrefrigerator? Or do you have anything else you want to fix before doing so?

personalizedrefrigerator commented:
@LittleHuba I think this is good to merge. However, to be clear, I have not applied [eliasrg@b9c50aa](https://github.com/eliasrg/xournalpp/commit/b9c50aab62ce58d157b5960f8d9c23dd1f98e7c4). Should I do this before merging? Should this be a part of the follow-up PR?

LittleHuba commented:
As this is essentially a bug in your code the fix should be included. Even for the small timeframe it will actually be active.

personalizedrefrigerator commented:
@LittleHuba Applied!

LittleHuba commented:
I'll go through the code once more tonight and make the functional test. If nothing major pops up I'll merge it then.

personalizedrefrigerator commented:
Testing the PR in KDE Plasma, Wayland, I see that the `setGtkTouchscreenScrollingEnabled`'s `Util::execInUiThread` with `G_PRIORITY_HIGH` is not called soon enough after a two-finger pan for the next stroke not to scroll. In other words, two-finger panning while drawing can cause single-finger gestures immediately after to scroll the view.

LittleHuba commented:
> Testing the PR in KDE Plasma, Wayland, I see that the `setGtkTouchscreenScrollingEnabled`'s `Util::execInUiThread` with `G_PRIORITY_HIGH` is not called soon enough after a two-finger pan for the next stroke not to scroll (at least, I think this is what's happening...). As such, two-finger panning while drawing sometimes causes the following single-finger gesture to scroll, rather than draw.
> Enabling the touch workaround fixes this, but this requires action by the user.

I was unable to reproduce this. Are you sure your finger actually leaves the screen inbetween? You can take a look at the event stream from the debug output to see if the events are sent.

The whole scrolling part seems to work fine. One problem I see is with the PDF caching. You should at least redraw once the zooming has stopped. Currently you only trigger redraws if a certain zoom delta is reached. This delta is chosen too high and leads to bad renders. Keep this logic but decrease the delta (or make it modifiable by the user) and then add a fixed redraw once zooming is finished and the delta is higher than some second very small delta. This way the user might have bad renders inbetween but a good one when he annotates again.

personalizedrefrigerator commented:
> You should at least redraw once the zooming has stopped. Currently you only trigger redraws if a certain zoom delta is reached. This delta is chosen too high and leads to bad renders.

Thank you for catching this!

personalizedrefrigerator commented:
@LittleHuba I've added a setting that controls how often pages are re-rendered. 

I think re-rendering the current page in the PDF when done zooming will require some refactoring. I haven't done this yet!

## Changed parts of the settings dialog
![image](https://user-images.githubusercontent.com/46334387/102540991-b2774400-4064-11eb-8657-1067461f102c.png)
![image](https://user-images.githubusercontent.com/46334387/102541051-c0c56000-4064-11eb-9f80-9a205ec76e65.png)

LittleHuba commented:
Thanks for the update! I'll retest the behavior tomorrow and get back to you with some feedback if necessary!

personalizedrefrigerator commented:
> I think re-rendering the current page in the PDF when done zooming will require some refactoring. I haven't done this yet!

I've been using the feature that limits PDF re-rendering (to the cached image) to edit large PDF files (~10 MB over 2 or 3 pages), for which a re-render can take a few seconds when zoomed in. Would leaving this setting as is be reasonable?

An alternative might be to replace it with a setting that prevents re-renders when the user has zoomed in more than a specified amount -- a greater zoom causes the PDFCache to render to a larger surface and therefore takes longer to render.

Febbe commented:
@LittleHuba any updates on this?

LittleHuba commented:
Sorry not yet. Currently busy with other stuff. I'll return to this PR shortly though.

Febbe commented:
The parts I had my fingers in, look very good. 
LGTM from me.

personalizedrefrigerator commented:
The most recent changes apply pressure inference at the start and end of strokes, addressing https://github.com/xournalpp/xournalpp/pull/2512#issuecomment-749090044.

LittleHuba commented:
Your changes have made a huge impact on zooming behaviour! There are two small things I'd like to change though.

With the current default value (1%) and the 5% steps you end up at 6%, 11% which is ugly. I do like the 5% steps though as they are crude enough for the normal user. People interested in fine adjustments can still type the value.

Instead I'd like the default value to be 5% as this keeps backgrounds sharp but reduces redrawing of the background by factor 5.

Once this is changed this PR should be good to go! Thanks for the good work!

personalizedrefrigerator commented:
Thank you for the feedback! I've changed this!

rolandlo commented:
I'm sorry, but with the latest merge of #2592  we have created some merge conflicts in the settings file. Could you please solve them?

personalizedrefrigerator commented:
@rolando I'm working on it!

## Additional changes

> Testing the PR in KDE Plasma, Wayland, I see that the `setGtkTouchscreenScrollingEnabled`'s `Util::execInUiThread` with `G_PRIORITY_HIGH` is not called soon enough after a two-finger pan for the next stroke not to scroll (at least, I think this is what's happening...).

I was able to reproduce this on GNOME Wayland. It was due to a mistake in `TouchInputHandler.cpp` (not properly setting `primarySequence` on touch end during multi-touch). I have fixed this.

In changing this, I also changed `gtkTouchscreenScrollingEnabled` to an `std::atomic_bool`, as it has potential to be passed between threads:
```c++
// Line 295
void MainWindow::setGtkTouchscreenScrollingEnabled(bool enabled) {
    if (enabled == gtkTouchscreenScrollingEnabled.load() || usingTouchWorkaround || winXournal == nullptr) {
        return;
    }

    gtkTouchscreenScrollingEnabled.store(enabled);

    Util::execInUiThread(
            [=]() {
                gtk_scrolled_window_set_kinetic_scrolling(GTK_SCROLLED_WINDOW(winXournal),
                                                          gtkTouchscreenScrollingEnabled.load());
            },
            G_PRIORITY_HIGH);
}
```

As such, `MainWindow.h` now includes `atomic`:
```c++
// Line 14
#include <atomic>
```

Febbe commented:
![grafik](https://user-images.githubusercontent.com/2228627/103180682-54cecd00-4898-11eb-935f-dc58df3c6539.png)
Both points require the restart also the point obviously requires also the touch workaround to be disabled not only the new input system.

Febbe commented:
> Zooming can be very slow when editing large PDFs.
>
>    Only partially fixes this issue

This may be fixed by #2577 or #2541

personalizedrefrigerator commented:
> ~Both points require the restart~ also the point obviously requires also the touch workaround to be disabled not only the new input system.

I'm confused... With the new input system and the workaround enabled, touch drawing should work. I think there's something I'm misunderstanding!

LittleHuba commented:
> > ~Both points require the restart~ also the point obviously requires also the touch workaround to be disabled not only the new input system.
> 
> I'm confused... With the new input system and the workaround enabled, touch drawing should work. I think there's something I'm misunderstanding!

You got it right @personalizedrefrigerator. Only the new input system requires the restart. Once it is enabled, switching this on and off functions without restart. In the long run we should remove these warnings from the descriptions and simply show the user a warning at the bottom of the settings dialog as soon as he has changed a setting that requires a restart.

@Febbe if you do not have anything additional I'd go ahead and merge this.

rolandlo commented:
@LittleHuba Does this PR need some squashing before it gets merged? It still has 12 commits.

LittleHuba commented:
Best would be to clean up some of the later commits by doing a Git history rewrite and hunk-based commits. That requires some higher knowledge of Git though and some planning in advance to know what change goes to what commit. Especially the fixup commits should be merged with the earlier ones.

If that is to much to request, we should just squash and merge it.

personalizedrefrigerator commented:
@LittleHuba I would rather have the commits squashed.

Febbe commented:
> I'm confused... With the new input system and the workaround enabled, touch drawing should work. I think there's something I'm misunderstanding!

The thing is, that the checkbox is checked, disabled but visible, when the Gtk-Touch-Workaround is checked. Only unchecking the Gtk-Touch-Workaround activates the checkbox.

personalizedrefrigerator commented:
> Only unchecking the Gtk-Touch-Workaround activates the checkbox. So the new input system is required to be checked and Gtk-Touch-Workaround must be unchecked. But that's not mentioned there.

The touch drawing setting has no effect on the old input system. The workaround, however, does. 

I'm not sure how to make this relationship clearer in the settings dialog...

LittleHuba commented:
@Febbe could you provide some information on how to improve this?

Febbe commented:
Maybe add a hint when the checkbox is disabled, that the Touch Workaround needs to be disabled to modify this option.

personalizedrefrigerator commented:
## Old input system enabled, workaround enabled

![image](https://user-images.githubusercontent.com/46334387/103445703-a37bbe80-4c2c-11eb-8f91-beb5efb4b8b5.png)

## Old input system enabled, workaround disabled

![image](https://user-images.githubusercontent.com/46334387/103445707-aeceea00-4c2c-11eb-924a-405b55f8f5f0.png)

## Old input system disabled, workaround enabled

![image](https://user-images.githubusercontent.com/46334387/103445725-dc1b9800-4c2c-11eb-842d-4c5bb67385e7.png)

## Old input system disabled, workaround disabled

![image](https://user-images.githubusercontent.com/46334387/103445729-ec337780-4c2c-11eb-8248-430e719d5a26.png)

Febbe commented:
Awesome 😊!

LittleHuba commented:
Thanks @personalizedrefrigerator for your hard work! I'm looking forward to future work of yours!

personalizedrefrigerator commented:
This was a rather large PR! Thank you for taking the time to review it!
